src/core/common/offset_range.hpp

```diff
@@ -0,0 +1,172 @@
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for an offset range.
+ */
+
+#ifndef OFFSET_RANGE_HPP_
+#define OFFSET_RANGE_HPP_
+
+#include "openthread-core-config.h"
+
+#include <openthread/platform/toolchain.h>
+
+#include "common/clearable.hpp"
+#include "common/num_utils.hpp"
+
+namespace ot {
+
+class Message;
+
+/**
+ * Represents an offset range.
+ *
+ */
+class OffsetRange : public Clearable<OffsetRange>
+{
+public:
+    /**
+     * Initializes the `OffsetRange`.
+     *
+     * @param[in] aOffset   The start offset.
+     * @param[in] aLength   The range length (number of bytes).
+     *
+     */
+    void Init(uint16_t aOffset, uint16_t aLength);
+
+    /**
+     * Initializes the `OffsetRange` from given start and end offsets.
+     *
+     * The range is inclusive of the start offset (@p aStartOffset) but exclusive of the end offset (@p aEndOffset).
+     *
+     * @param[in]  aStartOffset The start offset (inclusive).
+     * @param[in]  aEndOffset   The end offset (exclusive).
+     *
+     */
+    void InitFromRange(uint16_t aStartOffset, uint16_t aEndOffset);
+
+    /**
+     * Initializes the `OffsetRange` from a given `Message` from its offset to the message end.
+     *
+     * The start offset of the range is set to `aMessage.GetOffset()`, and the end offset is set to include all bytes
+     * in the message up to its current length `aMessage.GetLength()`.
+     *
+     * @param[in] aMessage    The `Message` to initialize the `OffsetRange` from.
+     *
+     */
+    void InitFromMessageOffsetToEnd(const Message &aMessage);
+
+    /**
+     * Initializes the `OffsetRange` from a given `Message` from zero offset up to to its full length.
+     *
+     * The start offset of the range is set to zero, and the end offset is set to include full length of @p aMessage.
+     *
+     * @param[in] aMessage    The `Message` to initialize the `OffsetRange` from.
+     *
+     */
+    void InitFromMessageFullLength(const Message &aMessage);
```

Comment:
This method seems only used in the unit test. Is it planned to be used in the future?

Reply:
Nice catch! Yes exactly. We will use this.

There are many other areas where we can use `OffsetRange` to simplify the code, e.g., in the `Coap` modules for processing options and blocks, or within the `ip6.cpp` etc. I have written some of these changes, but decided not to include them in this PR to keep it smaller and more focused (easier to review). Some of these are more substantial changes, so I plan submit these in separate follow-up PRs of their own.